core/schema/container.go

```diff
@@ -3,177 +3,500 @@ package schema
+			ArgDoc("source", `File to read the container from.`).
+			ArgDoc("tag", `Identifies the tag to import from the archive, if the archive bundles multiple tags.`),
+
+		dagql.Func("withRegistryAuth", s.withRegistryAuth).
+			Doc(`Retrieves this container with a registry authentication for a given address.`).
+			ArgDoc("address",
+				`Registry's address to bind the authentication to.`,
+				`Formatted as [host]/[user]/[repo]:[tag] (e.g. docker.io/dagger/dagger:main).`).
+			ArgDoc("username", `The username of the registry's account (e.g., "Dagger").`).
+			ArgDoc("secret", `The API key, password or token to authenticate to this registry.`),
+
+		dagql.Func("withoutRegistryAuth", s.withoutRegistryAuth).
+			Doc(`Retrieves this container without the registry authentication of a given address.`).
+			ArgDoc("address", `Registry's address to remove the authentication from.`,
+				`Formatted as [host]/[user]/[repo]:[tag] (e.g. docker.io/dagger/dagger:main).`),
+
+		dagql.Func("imageRef", s.imageRef).
+			Doc(`The unique image reference which can only be retrieved immediately after the 'Container.From' call.`),
+
+		dagql.Func("withExposedPort", s.withExposedPort).
+			Doc(`Expose a network port.`,
+				`Exposed ports serve two purposes:`,
+				`- For health checks and introspection, when running services`,
+				`- For setting the EXPOSE OCI field when publishing the container`).
+			ArgDoc("port", `Port number to expose`).
+			ArgDoc("protocol", `Transport layer network protocol`).
+			ArgDoc("description", `Optional port description`),
+
+		dagql.Func("withoutExposedPort", s.withoutExposedPort).
+			Doc(`Unexpose a previously exposed port.`).
+			ArgDoc("port", `Port number to unexpose`).
+			ArgDoc("protocol", `Port protocol to unexpose`),
+
+		dagql.Func("exposedPorts", s.exposedPorts).
+			Doc(`Retrieves the list of exposed ports.`,
+				`This includes ports already exposed by the image, even if not explicitly added with dagger.`),
+
+		dagql.Func("withServiceBinding", s.withServiceBinding).
+			Doc(`Establish a runtime dependency on a service.`,
+				`The service will be started automatically when needed and detached
+				when it is no longer needed, executing the default command if none is
+				set.`,
+				`The service will be reachable from the container via the provided hostname alias.`,
+				`The service dependency will also convey to any files or directories produced by the container.`).
+			ArgDoc("alias", `A name that can be used to reach the service from the container`).
+			ArgDoc("service", `Identifier of the service container`),
+
+		dagql.Func("withFocus", s.withFocus).
+			Doc(`Indicate that subsequent operations should be featured more prominently in the UI.`),
+
+		dagql.Func("withoutFocus", s.withoutFocus).
+			Doc(`Indicate that subsequent operations should not be featured more prominently in the UI.`,
+				`This is the initial state of all containers.`),
+
+		dagql.NodeFunc("shellEndpoint", s.shellEndpoint).Impure().
```

Comment:
Oh interesting - so to be clear, it's OK for them to re-use the same endpoint? The `identity.NewID()` is what looked like it shouldn't be cached.

In that case maybe we could/should just use a digest of the ID for the endpoint?

Reply:
Yeah I think I went with the random ID at the time because I was writing it and still wasn't sure what shape it was going to take, but all said and done it could easily just be a digest. This is all splitting hairs though because the cost of the implementation is just installing a handler, which is almost nothing, so very much a nit.

Reply:
Sounds good, just switched this to pure and changed it to use a digest :+1: